enhancements/art/flexibly-enforcing-product-build-constraints-in-ci.md

```diff
@@ -0,0 +1,268 @@
+
+### Goals
+The current alignment implementation, prior to this enhancement, had the following goals:
+- The ART pipelines will provide up-to-date, downstream equivalent images within the CI environment. 
+  These will be referred to as "ART equivalent" images.
+- Downstream metadata describing how to build production images which compose a release will be 
+  analyzed automatically and used to open upstream PRs for Dockerfiles when those Dockerfiles 
+  are not aligned with the current state of the production metadata's ART equivalent images.
+
+Enhancement Goal:
+- If a component's upstream Dockerfile is using ART equivalent images, the downstream build process 
+  will honor the builder images within that Dockerfile until shortly before feature freeze. This 
+  will provide upstream teams:
+    - Visibility that a change is being recommended.
+    - The time to test those changes with full test coverage (not only openshift-tests periodics, 
+      but also unit, integration, operator tests etc.).
+    - The space to align their production metadata if the change is not appropriate for their component.
+  All before production builds begin to use the suggested change.
+- Ensure that all alignment PRs are either merged or closed (after discussion with ART to change 
+  production metadata) well before GA.
+- Prevent long term divergence between upstream and downstream builds.
+- Prevent components from lingering on aging golang versions indefinitely. Older versions of 
+  golang carry a growing support cost and security risk that is better addressed by teams 
+  migrating to new versions.
+- CI and downstream ART builds use the same image specification in the upstream repositories 
+  until feature freeze. Disparities between CI and ART after this point will be visible and 
+  tracked (first by PRs and then by BZs if the disparity remains).
+- To make evident, through the release controller, when a release is inconsistent 
+  between upstream and downstream in order to provide organizational visibility. 
+
+### Non-Goals
+- Improve coverage of minor Golang updates of the platform. Those updates only run openshift-tests 
+  based tests (promotion jobs). We are aware that minor Golang updates do NOT have full test coverage, 
+  e.g. unit, integration, operator tests are NOT run.
+- To allow upstream repositories to use arbitrary builder images. Even if this were desirable, 
+  it cannot be reproduced behind the firewall -- production builds cannot access the Internet 
+  and all inputs to a given production build must themselves be produced as part of a production 
+  build (e.g. an arbitrary golang version chosen upstream cannot be used downstream until a Red 
+  Hat internal team builds  -- and supports -- that golang version as part of a production build).
+- Within a given z-stream, allow users to pick a minor (eg. 1.15.1 vs 1.15.2) version of golang. 
+  ART ensures that the latest available builder image provided by the golang team is used.
```

Comment:
This statement goes against the previous, unless the example is wrong.

Reply:
Changing to `Instead, this proposal will ensure that the latest available patch level of the golang builder image provided by the Red Hat golang team is used.`